internal/repository/indexmap.go

```diff
@@ -0,0 +1,140 @@
+package repository
+
+import (
+	"encoding/binary"
+
+	"github.com/restic/restic/internal/restic"
+)
+
+// An indexMap is a chained hash table that maps blob IDs to indexEntries.
+// It allows storing multiple entries with the same key.
+//
+// A chained table has less memory overhead than a probing one (like the
+// standard map), because only a pointer array needs to be resized when the
+// table grows.
+type indexMap struct {
+	// The number of buckets is always a power of two and never zero.
+	buckets    []*indexEntry
+	numentries uint
+
+	free *indexEntry // Free list.
+}
+
+const (
+	growthFactor = 2 // Must be a power of 2.
+	maxLoad      = 4 // Max. number of entries per bucket.
+)
+
+// add inserts an indexEntry for the given arguments into the map,
+// using id as the key.
+func (m *indexMap) add(id restic.ID, packIdx int, offset, length uint32) {
+	switch {
+	case m.numentries == 0: // Lazy initialization.
+		m.buckets = make([]*indexEntry, 64)
+	case m.numentries >= maxLoad*uint(len(m.buckets)):
+		m.grow()
+	}
+
+	h := m.hash(id)
+	e := m.newEntry()
+	e.id = id
+	e.next = m.buckets[h]
+	e.packIndex = packIdx
+	e.offset = offset
+	e.length = length
+
+	m.buckets[h] = e
+	m.numentries++
+}
+
+// foreach calls fn for all entries in the map, until fn returns false.
+func (m *indexMap) foreach(fn func(*indexEntry) bool) {
+	for _, e := range m.buckets {
+		for e != nil {
+			if !fn(e) {
+				return
+			}
+			e = e.next
+		}
+	}
+}
+
+// foreachWithID calls fn for all entries with the given id.
+func (m *indexMap) foreachWithID(id restic.ID, fn func(*indexEntry)) {
+	if len(m.buckets) == 0 {
+		return
+	}
+
+	h := m.hash(id)
+	for e := m.buckets[h]; e != nil; e = e.next {
+		if e.id != id {
+			continue
+		}
+		fn(e)
+	}
+}
+
+// get returns the first entry for the given id.
+func (m *indexMap) get(id restic.ID) *indexEntry {
+	if len(m.buckets) == 0 {
+		return nil
+	}
+
+	h := m.hash(id)
+	for e := m.buckets[h]; e != nil; e = e.next {
+		if e.id == id {
+			return e
+		}
+	}
+	return nil
+}
+
+func (m *indexMap) grow() {
+	old := m.buckets
+	m.buckets = make([]*indexEntry, growthFactor*len(m.buckets))
+
+	for _, e := range old {
```

Comment:
Maybe we should consider to run this loop concurrently? Within one iteration, only buckets for new hashes are accessed and these are completely isolated when using a growFactor of two. I'm just fearing that growing large indexes will keep only 1 CPU occupied for quite some time...

Reply:
You mean like so?

```go
var sem = make(chan struct{}, runtime.NumCPU())

func (m *indexMap) grow() {
	old := m.buckets
	m.buckets = make([]*indexEntry, growthFactor*len(m.buckets))

	if len(m.buckets) < 8192 {
		for _, e := range old {
			for e != nil {
				h := m.hash(e.id)
				next := e.next
				e.next = m.buckets[h]
				m.buckets[h] = e
				e = next
			}
		}
		return
	}

	for _, e := range old {
		m.wg.Add(1)
		sem <- struct{}{}

		go func(e *indexEntry) {
			for e != nil {
				h := m.hash(e.id)
				next := e.next
				e.next = m.buckets[h]
				m.buckets[h] = e
				e = next
			}

			<-sem
			m.wg.Done()
		}(e)
	}

	m.wg.Wait()
	m.wg = sync.WaitGroup{}
}
```

This does:

```
name          old time/op    new time/op    delta
DecodeIndex-8    16.1µs ± 4%    16.0µs ± 3%     ~     (p=0.571 n=12+10)
IndexAlloc-8      629ms ± 2%     827ms ± 1%  +31.55%  (p=0.000 n=11+10)
```

I can see it using at most 3.5 of my 8 cores in top. I suppose it's too memory-bound.

We could try a preallocated goroutine pool, but this function is mostly called from worker pools doing SaveBlob or DecodeIndex anyway.

Reply:
A nil check before ch <- e helps a bit, but not much.

```
name           old time/op    new time/op    delta
DecodeIndex-8    16.1µs ± 4%    15.8µs ± 3%     ~     (p=0.143 n=12+9)
IndexAlloc-8      629ms ± 2%     803ms ± 2%  +27.76%  (p=0.000 n=11+10)
```

Reply:
Too much of communication :flushed: 

Using
```
cpus := runtime.NumCPU()
	var wg sync.WaitGroup
	wg.Add(cpus)
	for cpu := 0; cpu < cpus; cpu++ {
		go func(cpu int) {
			for i := cpu; i < len(m.buckets); i += cpus {
				b := m.buckets[i]
				for b != nil {
					index := b.key.Hash() & m.mask
					next := b.next
					b.next = newBuckets[index]
					newBuckets[index] = b
					b = next
				}
			}
			wg.Done()
		}(cpu)
	}
	wg.Wait()
```
in my code (but should be equal to your implementation) gives me:
```
alex@thinkpad:~/go/src/github.com/restic/restic/internal/repository$ benchstat old new
name          old time/op    new time/op    delta
IndexAlloc-4     1.06s ± 9%     0.90s ± 3%  -15.34%  (p=0.000 n=10+10)

name          old alloc/op   new alloc/op   delta
IndexAlloc-4     410MB ± 0%     410MB ± 0%     ~     (p=0.315 n=10+10)

name          old allocs/op  new allocs/op  delta
IndexAlloc-4     2.63M ± 0%     2.63M ± 0%     ~     (p=0.156 n=10+9)
```

Reply:
Nice! What size cap did you use to decide on sequential vs. parallel?

Reply:
In fact I didn't use any. My judgement is that optimizing growing for small tables is not worth to optimize for, but I didn't benchmark it, though.

Reply:
7897775f64ff3a31e36adc301a94938791294c3e.